### PR TITLE
(Feature) Account Management

### PR DESCRIPTION
**Pull request by sfount**
This pull request introduces the 2.x designed Accounts Management module as proposed in https://github.com/IMA-WorldHealth/bhima-2.X/issues/489 following the design https://github.com/IMA-WorldHealth/bhima-2.X/issues/269.

**Key Features**
- Account routes are defined on the client in a seperate `accounts.route.js` found in the Accounts partials (module) folder as proposed in https://github.com/IMA-WorldHealth/bhima-2.X/issues/494. This can act as a proof of concept for cleaning up the client `app.js` routes.
- State change management for modals - modals are now dismissed on state change using the `$uibModal.dismissAll()` method on state change start event. This along with designed UI router rules allows for a clean modal state implementation that respects the browsers back button. 

`/accounts` - Account management page 
`/accounts/:id` - Selects an individual account (can be implemented if required)
`/accounts/:id/edit` - Enter update state for account with 'id'
`/accounts/create` - Enter create state - if a parent id is passed in during the state change, this will automatically populate the parent field

**Technical Features**
- Introduces `bhConstant`, currently this defines a JSON object in `app.js` that can be injected and used throughout the application, as defined https://github.com/IMA-WorldHealth/bhima-2.X/issues/460
- Account schema change to prepare for account tags/ report structure
- UI Grid treeview implementation with nested account relationships
- Translatable account type keys
- Minor modal style update

**Module Overview**

_Accounts View - Respects Nested Account Relationships_
screenshot 2016-06-28 at 10 01 54

_Title Accounts Expandable/ Collapsible - Feature abstracted for use in any module_
screenshot 2016-06-28 at 10 08 43
screenshot 2016-06-28 at 10 08 31

_Create State - /accounts/create_
screenshot 2016-06-28 at 10 02 31

_Update State - /accounts/:id/edit_
- _Client only supports updating label and parent fields_
- _Root account selected description shown for root account selection_
  screenshot 2016-06-28 at 10 04 35

_Batch account create feature - resets modal on account creation remembering parent and account type selection_
screenshot 2016-06-28 at 10 04 05

_Handle incorrect or outdated account edit link_
screenshot 2016-06-28 at 11 20 57

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/489 

---
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

### CODE REVIEW

File: client/src/js/app.js [hunk 1]
```diff
@@ -550,6 +548,19 @@ function localStorageConfig($localStorageProvider) {
 }
 
 /**
+ * @todo some of these constants are system standards, others should be
+ * populated according to the enterprise configuration
+ */
+function constantConfig() {
```
Comment (jniles) [inline client/src/js/app.js:554]: :+1:

File: client/src/js/app.js [hunk 2]
```diff
@@ -522,6 +516,10 @@ function startupConfig($rootScope, $state, SessionService, amMoment, Notify, $lo
       Notify.warn('AUTH.CANNOT_RETURN_TO_LOGIN');
     }
 
+    // clean up any modals that are currently active, this allows modules to use onEnter state modals
+    // without considering all exit cases
+    $uibModalStack.dismissAll();
```
Comment (jniles) [inline client/src/js/app.js:521]: :dancer:

File: client/src/js/app.js [hunk 3]
```diff
@@ -1,17 +1,16 @@
 var bhima = angular.module('bhima', [
   'bhima.controllers', 'bhima.services', 'bhima.directives', 'bhima.filters',
-  'bhima.components', 'ui.router', 'ui.bootstrap', 'pascalprecht.translate',
+  'bhima.components', 'bhima.routes', 'ui.router', 'ui.bootstrap', 'pascalprecht.translate',
```
Comment (jniles) [inline client/src/js/app.js]: Do we still need `ui.router` here if it is in the `bhima.routes`?

File: client/src/js/services/Store.js [hunk 4]
```diff
@@ -63,7 +63,7 @@ function StoreService() {
    * @method post
    *
    * @description
-   * This method inserts and object into the store.  If the object is missing
+   * This method inserts an object into the store.  If the object is missing
```
Comment (jniles) [inline client/src/js/services/Store.js:66]: :dancer:

File: client/src/js/services/AccountService.js [hunk 5]
```diff
@@ -138,13 +143,18 @@ function AccountService($http, util, Session) {
    * This Method Creat an account
    */
   function create(account) {
-    var classAccount = String(account.number).charAt(0),
+    
+    if (!angular.isDefined(account.enterprise_id)) {
```
Comment (jniles) [inline client/src/js/services/AccountService.js]: Would it be better to do this on the server?

File: client/src/js/services/AccountService.js [hunk 6]
```diff
@@ -138,13 +143,18 @@ function AccountService($http, util, Session) {
    * This Method Creat an account
    */
   function create(account) {
-    var classAccount = String(account.number).charAt(0),
+    
+    if (!angular.isDefined(account.enterprise_id)) { 
+      account.enterprise_id = Session.enterprise.id;
+    }
+    /*var classAccount = String(account.number).charAt(0),
```
Comment (jniles) [inline client/src/js/services/AccountService.js]: Go ahead and remove these commented blocks.  They won't be used anymore.

File: client/src/js/services/util.js [hunk 7]
```diff
@@ -21,15 +22,16 @@ function UtilService($filter, moment) {
...
-
-          // accounts for empty string values
-          response[key] = angular.isDefined(value.$modelValue) ? value.$modelValue : value.$bhValue;
+        if (requireDirty && !value.$dirty) {
+          return;
         }
+        
+        // any standard ng-model element will provide an $modelValue, according
+        // to the latest 2.x standards more complex bhima components or bhima
+        // component wrappers will expose $bhValue
```
Comment (jniles) [inline client/src/js/services/util.js:31]: The $bhValue should probably be well documented. :)  I had forgotten about this myself.

File: client/src/less/bhima-bootstrap.less [hunk 8]
```diff
@@ -56,6 +56,12 @@ growl-notifications growl-notification {
   width : calc(~'100% - @{collapsed-offset}');
 }
 
+@modal-title-padding : 8px;
+@modal-inner-padding : 10px;
```
Comment (jniles) [inline client/src/less/bhima-bootstrap.less:60]: :+1:

File: client/src/partials/accounts/AccountGrid.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+ * @description
+ * This service is responsible for collecting data required to populate the
+ * Account Management module, it also provides helper methods for dynamically
+ * adding and removing data.
+ */
+function AccountGridService(AccountStore, Accounts, Store, Notify) {
+  /**
+   * @constructor
+   *
+   * @description
```
Comment (jniles) [inline client/src/partials/accounts/AccountGrid.js:18]: Nice comments!

File: client/src/partials/accounts/edit/accounts.edit.js [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+  vm.state = angular.copy($state.current.name);
+  vm.isCreateState = vm.state === vm.states.create;
+
+  // varaibles to track custom modal error handling, these will be replaced
+  // with either the Notification library or a uniform modal error handling utility
+  vm.fetchError = null;
+  vm.accountNotFound = null;
+
+  vm.rootAccount = {
```
Comment (jniles) [inline client/src/partials/accounts/edit/accounts.edit.js:40]: I think `rootAccount` might go in the `bhConstants`.  What do you think?

File: client/src/partials/accounts/edit/accounts.edit.js [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+  }
+
+  /**
+   * @method mockAccountNotFound
+   *
+   * @description
+   * This method mocks a 404 returned from the database - if this becomes a common
+   * pattern on the client this could be handled and returned from the store
+   */
+  function mockAccountNotFound() {
```
Comment (jniles) [inline client/src/partials/accounts/edit/accounts.edit.js:218]: :sake:

**Comment by jniles** [comment]
@sfount, I've done an initial code review.

This page looks good to me.  I'm made a few comments, mostly related to cleanup of comments and old code.  Take a look and let me know what you think.

This can be merged anytime.

File: client/src/partials/accounts/edit/accounts.edit.js (hunk 10 above)
Comment (sfount) [inline client/src/partials/accounts/edit/accounts.edit.js:40]: The only place I can think of using the root account right now is when creating accounts - the rest of the time it is just implicit. 

I think if it needs to be used in more than one place this is a good idea.

File: client/src/js/app.js (hunk 3 above)
Comment (sfount) [inline client/src/js/app.js]: Good catch!

File: client/src/partials/accounts/edit/accounts.edit.js (hunk 10 above)
Comment (jniles) [inline client/src/partials/accounts/edit/accounts.edit.js:40]: Okay, sounds good.

**Comment by sfount** [comment]
@jniles Updated with your feedback
- Removed commented legacy cleaning blocks 
- Assign enterprise ID on the server 
- Remove redundant import of `ui.router`

**Comment by jniles** [comment]
LGTM.  As soon as tests pass, this will be merged.

**Pull request merged**